lib/msal-common/src/client/AuthorizationCodeClient.ts

```diff
@@ -0,0 +1,185 @@
+
+        const authority: Authority = await this.createAuthority(request && request.authority);
+        const queryParams: Map<string, string> = this.generateAuthCodeUrlParams(request);
+        const queryString = RequestUtils.createQueryString(queryParams);
+        return authority.authorizationEndpoint + "?" + queryString;
+    }
+
+    /**
+     * API to acquire a token in exchange of 'authorization_code` acquired by the user in the first leg of the authorization_code_grant
+     * @param request
+     */
+    async acquireToken(request: AuthorizationCodeRequest): Promise<string> {
+
+        const authority: Authority = await this.createAuthority(request && request.authority);
+        const acquiredTokenResponse = this.executeTokenRequest(authority, request);
+        return acquiredTokenResponse;
+
+        // add response_handler here to send the response
+    }
+
+    /**
+     * Executes POST request to token endpoint
+     * @param authority
+     * @param request
+     */
+    private async executeTokenRequest(authority: Authority, request: AuthorizationCodeRequest): Promise<string> {
+
+        const tokenParameters: Map<string, string> = this.generateAuthCodeParams(request);
+        const requestBody = RequestUtils.createQueryString(tokenParameters);
+        const headers: Map<string, string> = this.createDefaultTokenRequestHeaders();
+
+        let acquiredTokenResponse;
+        try {
+            acquiredTokenResponse = this.executePostToTokenEndpoint(
+                authority.tokenEndpoint,
+                requestBody,
+                headers);
+
+            return acquiredTokenResponse;
+        } catch (error) {
+            console.log(error.response.data); // TODO use logger
+            return error.response.data;
+        }
+    }
+
+    /**
+     * Generates a map for all the params to be sent to the service
+     * @param request
+     */
+    private generateAuthCodeParams(request: AuthorizationCodeRequest) : Map<string, string> {
+        const paramsMap: Map<string, string> = new Map<string, string>();
+
+        RequestUtils.addClientId(paramsMap, this.config.authOptions.clientId);
+
+        // validate and add scopes
+        const scopes = RequestValidator.validateAndGenerateScopes(
+            request.scopes,
+            this.config.authOptions.clientId
+        );
+        RequestUtils.addScopes(paramsMap, scopes);
+
+        // validate the redirectUri (to be a non null value)
+        RequestValidator.validateRedirectUri(request.redirectUri);
+        RequestUtils.addRedirectUri(paramsMap, request.redirectUri);
+
+        // add code: user set, not validated
+        RequestUtils.addAuthorizationCode(paramsMap, request.code);
+
+        // add code_verifier if passed
+        if (request.codeVerifier) {
+            RequestUtils.addCodeVerifier(paramsMap, request.codeVerifier);
+        }
+
+        RequestUtils.addGrantType(paramsMap, GrantType.AUTHORIZATION_CODE_GRANT);
```

Comment:
instead of side effects is there a way we can do something like this 
```js
const paramsMap: Map<string, string> RequestUtils.generateParamsMapFromRequest):
```

then in  the utils you can do all your logic around setting int, but writing functions that take maps and just "alter in place" can create bugs and is harder to test.